src/lib.rs

```diff
@@ -1,3 +1,138 @@
+//! A proxy actor and infrastructure for asynchronous and clear cache interaction
+//!
+//! # A tour of actix-cache
+//!
+//! This crate consist of three main part:
+//! * [Cache] actor.
+//! * [Backend] trait and its implementation ([RedisBackend]).
+//! * [Cacheable] trait.
+//!
+//! ## Features
+//! * Async/Sync cache backend support.
+//! * Dogpile effect prevention.
+//! * Stale cache mechanics.
+//! * Automatic cache key generation.
+//! * Detailed Prometheus metrics out of the box.
+//!
+//! ## Feature flags
+//! * derive - Support for [Cacheable] trait derive macros.
+//! * metrics - Support for Prometheus metrics.
+//!
+//! ## Example
+//! First of all, you should derive [Cacheable] trait for your actix Message:
+//!
+//! > **_NOTE:_** Default cache key implementation based on serde_qs crate
+//! > and have some [restrictions](https://docs.rs/serde_qs/latest/serde_qs/#supported-types). 
+//!
+//!
+//! ```rust
+//! use actix::prelude::*;
+//! use actix_cache::Cacheable; // With features=["derive"]
+//! use actix_derive::Message;
+//! use serde::{Deserialize, Serialize};
+//! # struct Pong;
+//!
+//! #[derive(Message, Cacheable, Serialize)]
+//! #[rtype(result = "Result<Pong, ()>")]
+//! struct Ping {
+//!     id: i32,
+//! }
+//! ```
+//! Or implement that trait manually:
+//!
+//! ```rust
+//! # use actix_cache::{Cacheable, CacheError};
+//! # struct Ping { id: i32 }
+//! impl Cacheable for Ping {
+//!     fn cache_key(&self) -> Result<String, CacheError> {
+//!         Ok(format!("Ping::{}", self.id))
+//!     }
+//! }
+//! ```
+//! Next step is to instantiate [Cache] actor with selected backend:
+//!
+//! ```rust
+//! # use actix::prelude::*;
+//! use actix_cache::{CacheError, Cache as CacheActor, RedisBackend};
+//!
+//! type Cache = CacheActor<RedisBackend>;
+//!
+//! #[actix_rt::main]
+//! async fn main() -> Result<(), CacheError> {
+//!     let cache = Cache::new()
+//!         .await?
+//!         .start();
+//! #   Ok(())
+//! }
+//! ```
+//!
+//! And the last step is using cache in your code (actix-web handler for example):
+//!
+//! ```rust
+//! # use actix::prelude::*;
+//! # use actix_web::{web, App, HttpResponse, HttpServer, Responder};
+//! # use actix_cache::{Cache as CacheActor, RedisBackend, Cacheable};
+//! # use serde::Serialize;
+//! #
+//! # struct FibonacciActor;
+//! #
+//! # impl Actor for FibonacciActor { type Context = Context<Self>; }
+//! #
+//! # #[derive(Message, Cacheable, Serialize)]
+//! # #[rtype(result = "u64")]
+//! # struct GetNumber {
+//! #     number: u8
+//! # }
+//! #
+//! # impl Handler<GetNumber> for FibonacciActor {
+//! #     type Result = <GetNumber as Message>::Result;
+//! #
+//! #     fn handle(&mut self, msg: GetNumber, _ctx: &mut Self::Context) -> Self::Result {
+//! #         42
+//! #     }
+//! # }
+//! #
+//! # type Cache = CacheActor<RedisBackend>;
+//! async fn index(
+//!     fib: web::Data<Addr<FibonacciActor>>,
```

Comment:
What do you think about this example?
```rust
async fn index(
    db: web::Data<Addr<DatabaseActor>>,
    cache: web::Data<Addr<Cache>>
) -> impl Responder {
    let query = GetPosts { limit: 10 };
    let posts = cache
        .send(query.into_cache(&db))
        .await
        .unwrap()
        .unwrap();
    HttpResponse::Ok().body(format!("There are {} posts", posts.len()))
}
```

Reply:
Nice. But this example will not be built without a declaration of all necessary types. And can't be a doctest.

Reply:
I think we should add a separate example for actix-web + diesel. But it must be a different repository. I don't want to add diesel into dev-dependencies.

Reply:
#14